src/passes/GUFA.cpp

```diff
@@ -0,0 +1,308 @@
+    }
+    return !EffectAnalyzer(getPassOptions(), *getModule(), curr)
+              .hasUnremovableSideEffects();
+  }
+
+  // Whether we can replace something (but not its children, we can keep them
+  // with drops) with an unreachable without changing observable behavior or
+  // breaking validation.
+  bool canReplaceWithUnreachable(Expression* curr) {
+    if (!canRemoveStructurally(curr)) {
+      return false;
+    }
+    EffectAnalyzer effects(getPassOptions(), *getModule(), curr);
+    // Ignore a trap, as the unreachable replacement would trap too.
+    effects.trap = false;
+    return !effects.hasUnremovableSideEffects();
+  }
+
+  void visitExpression(Expression* curr) {
+    // Skip things we can't improve in any way.
+    auto type = curr->type;
+    if (type == Type::unreachable || type == Type::none ||
+        Properties::isConstantExpression(curr)) {
+      return;
+    }
+
+    if (type.isTuple()) {
+      // TODO: tuple types.
+      return;
+    }
+
+    if (type.isRef() && (getTypeSystem() != TypeSystem::Nominal &&
+                         getTypeSystem() != TypeSystem::Isorecursive)) {
+      // Without type info we can't analyze subtypes, so we cannot infer
+      // anything about refs.
+      return;
+    }
+
+    // Ok, this is an interesting location that we might optimize. See what the
+    // oracle says is possible there.
+    auto contents = oracle.getContents(ExpressionLocation{curr, 0});
+
+    auto& options = getPassOptions();
+    auto& wasm = *getModule();
+    Builder builder(wasm);
+
+    auto replaceWithUnreachable = [&]() {
+      if (canReplaceWithUnreachable(curr)) {
+        replaceCurrent(getDroppedChildrenAndAppend(
+          curr, wasm, options, builder.makeUnreachable()));
+      } else {
+        // We can't remove this, but we can at least put an unreachable
+        // right after it.
+        replaceCurrent(builder.makeSequence(builder.makeDrop(curr),
+                                            builder.makeUnreachable()));
+      }
+      optimized = true;
+    };
+
+    if (contents.getType() == Type::unreachable) {
+      // This cannot contain any possible value at all. It must be unreachable
+      // code.
+      replaceWithUnreachable();
+      return;
+    }
+
+    // This is reachable. Check if we can emit something optimized for it.
+    // TODO: can we handle more general things here too?
+    if (!contents.canMakeExpression()) {
+      return;
+    }
+
+    if (contents.isNull() && curr->type.isNullable()) {
+      // Null values are all identical, so just fix up the type here (the null's
+      // type might not fit in this expression).
+      contents =
+        PossibleContents::literal(Literal::makeNull(curr->type.getHeapType()));
+
+      // Note that if curr's type is *not* nullable, then the code will trap at
+      // runtime (the null must arrive through a cast that will trap). We handle
+      // that below, so we don't need to think about it here.
```

Comment:
When can this happen? Does that validate then?

Reply:
Say,
```wat
(block (result (ref $A)) ;; non-nullable
  (ref.as_non_null
    (ref.null $A)))
```
That validates but will trap at runtime.

Reply:
Thanks, I think some of my questions effectively asked the same thing. I didn't think about cast instructions so I was assuming `ref.null` is the only instruction that can contain a null value and confused.